Manage cluster credit balance

jamesremuscat opened the pull request:
This PR (re-)introduces the concept of a per-cluster support credit balance, and adds features to support its use and management.

Admins may now add `CreditDeposit`s to a `Cluster` to increase its credit balance. This balance is reduced by charges on closed `Case`s, but can be zero or negative without error or restriction.

The current balance for a `Cluster` is shown on the newly-remodelled cluster dashboard, and a new page is introduced to allow users (both admins and not) to inspect detail of credit usage per quarter. Each closed case (regardless of charge amount, even if zero) and credit deposit is listed in this fashion, and totals for credits accrued, credits used, and cases closed with zero charge are displayed.

Closes #296.

Trello: https://trello.com/c/8HVAKl8W/332-credit-usage-summary

bobwhitelock commented on app/controllers/cases_controller.rb:
```diff
@@ -68,7 +68,7 @@ def create
   def close
     charge = params.require(:case).require(:credit_charge).to_i
```
Also, currently if I attempt to close a Case with a negative number of credits I currently get the following error in the console:

  ```
  Uncaught TypeError: Cannot read property 'style' of undefined
  ```

My (possibly incorrect) guess is this has something to do with Rails' unobtrusive JS trying to disable the submit button when the form is submitted, and this not working in this case for some reason.

I think we do want to disable closing a case with a negative number of credits, as this doesn't seem a meaningful thing to do, but this should probably be done by preventing this client-side/ a validation server-side (will discuss this more below).

jamesremuscat replied:
FWIW I can't reproduce that `TypeError`.

There's already HTML5 validation that this amount is non-negative, which for me prevents submitting the form with negative values (though it does so quietly and after the Rails `data-confirm` popup).

bobwhitelock replied:
Apologies, the `TypeError` was apparently due to some badly written Chrome extension I have getting confused at the point where I submit the form, which made me think it was related. I do actually get that same behaviour (a quiet failure after the `data-confirm` popup) - I guess this is slightly confusing the way that this happens (which seems to be a Rails + HTML5 validation thing?), but not a major issue as this is an admin-only form.

One other thing I have now noticed related to this: if I close a Case and enter nothing in the `amount` field (just deleting the default `value`) then the Case will be closed without a CreditCharge created as `amount` is blank. This seems like not what we want to happen to me as I believe we want the credit usage page to give a comprehensive list of the credit usage for all closed Cases, and I believe this means closed Cases could be hidden from this list. Some suggestions related to this:

1. We probably want to use `create_credit_charge!` here: https://github.com/alces-software/alces-flight-center/blob/f9d709653bf84d08a92b2f5b98b2697717b7e098/app/controllers/cases_controller.rb#L79-L80 since we always want this to happen. Aside: when this does happen it seems to make the saved Case invalid, due to this validation: https://github.com/alces-software/alces-flight-center/blob/f9d709653bf84d08a92b2f5b98b2697717b7e098/app/models/case.rb#L74

    however the controller is still able to call `kase.close!` successfully as at the point this is called the Case instance we have is valid as it does have a `credit_charge`, this just hasn't been saved and is invalid.

2. It might be worth adding `required` to the input to prevent the submit occurring in this situation, since it is required that an `amount` for the CreditCharge be given.

jamesremuscat replied:
Thinking about it, I decided to add a `validates_associated :credit_charge` to `Case`, since this means a sensible error gets added to the flash.

bobwhitelock commented on app/models/credit_deposit.rb:
```diff
@@ -0,0 +1,5 @@
+  belongs_to :cluster
+  delegate :site, to: :cluster
+  attr_readonly :cluster
+end
```
I've briefly mentioned this elsewhere, but to discuss this just in one place - we have the requirement that we want to allow the overall credit balance to take any value at the moment. However I think it would still make sense to validate the `amount` for both `CreditCharge`s and `CreditDeposit`s:

- for `CreditCharge`s, I think we should ensure `amount >= 0`, I'm not sure that adding credits to a Cluster when a Case is closed will be meaningful to Users or as part of the current business process;

- for `CreditDeposit`s, I think we should ensure `amount > 0`, similarly I'm not sure making credit deposits which actually reduce the number of credits for a Cluster makes sense, and depositing 0 credits doesn't seem a useful thing to support and is likely to confuse Users.

When we have these validations it would then also make sense to have corresponding HTML5 validations on the inputs as well.

jamesremuscat replied:
I was being deliberately 'light-touch' with the validation around credit amounts - perhaps a little too so, as clearly it's at best confusing to have a negative-valued credit charge associated with a case.

For deposits - the sort-of rough use case I had in mind was if we ever had to issue a correction (e.g. someone puts an extra 0 in by mistake when adding a deposit) but perhaps it's best to wait and only support that use case properly if it's needed.

jamesremuscat commented on config/routes.rb:
```diff
@@ -140,6 +141,7 @@
       logs.call
       confirm_maintenance_form.call
       get :documents
+      get '/credit_usage(/:start_date)', to: 'clusters#credit_usage', as: :credit_usage
```
It's possible my Rails-fu is insufficient but I don't see how to define the `:start_date` param using `resources`.

I suppose we could have less-pretty URLs by requiring it in the non-route-param way (e.g. `/credit-usage?start_date=2018-01-01` instead of `/credit-usage/2018-01-01`) if you really wanted?

jamesremuscat replied:
Reading it back, that last comment seemed far more passive-aggressive than I'd intended - sorry!

I've changed the URL to `/credit-usage` but left it as a `get` definition rather than a `resource`.

bobwhitelock replied:
> Reading it back, that last comment seemed far more passive-aggressive than I'd intended - sorry!

No worries, I didn't read it that way! But I agree conveying sentiment in text is hard :).

Looking at it myself how to do this also doesn't immediately jump out to me, so I think this is fine as it is, it would be a pretty minor difference anyway.

One minor other comment re:

> I suppose we could have less-pretty URLs by requiring it in the non-route-param way (e.g. `/credit-usage?start_date=2018-01-01` instead of `/credit-usage/2018-01-01`)

this is currently what the URLs look like to me, at least when I navigate on a credit usage page via the select box, e.g. `http://center.alces-flight.lvh.me:3000/clusters/1/credit-usage?start_date=2018-04-01` - is this not what you intended?

jamesremuscat replied:
No, that's not what I intended - clearly wasn't paying attention to the URL when I wrote that part! I've just changed the JavaScript that triggers page navigation from that `select` so that it uses pretty URLs. JavaScript-less people will still end up with the long-form URL when they click "View".

jamesremuscat commented on app/controllers/clusters_controller.rb:
```diff
@@ -1,3 +1,71 @@
+
+    @used = add_up(charges)
+
+    @free_of_charge = charges.where(amount: 0).count
```
I moved it to `Cluster`, then decided it was too messy having it there so broke out all the cluster-charge-events-in-given-period stuff into its own class.

bobwhitelock replied:
Yep that looks great, thanks

jamesremuscat commented on app/controllers/clusters_controller.rb:
```diff
@@ -1,3 +1,71 @@
 class ClustersController < ApplicationController
```
I went ahead and did this, it was straightforward.

bobwhitelock replied:
Thanks :slightly_smiling_face:

jamesremuscat commented:
Rebased on latest `develop` to solve conflicts.

bobwhitelock commented:
Is there anything still to add to this @jamesremuscat or do you want me to take another look? :slightly_smiling_face:

jamesremuscat commented:
There was a comment for you to consider about the credits URL, I've addressed all the other things needing addressing. :slightly_smiling_face:

bobwhitelock commented on db/data/20180611161221_migrate_legacy_case_charges.rb:
```diff
@@ -1,9 +1,13 @@
   def up
     Case.where(state: 'closed').each do |k|
       unless k.credit_charge.present?
         k.create_credit_charge(amount: k.legacy_credit_charge)
```
Now that this migration has been switched to the new format I've tried running it (via `rake alces:db:import_and_migrate_production`) - however when I did this I noticed no `CreditCharge`s were created, although there are currently two closed Cases. If I change this line to use the bang method (`create_credit_charge!`) it indicates why, due to a failing validation:

```
StandardError: An error has occurred, this and all later migrations canceled:

Validation failed: User must exist
/home/bob/src/alces-software/alces-flight-center/db/data/20180611161221_migrate_legacy_case_charges.rb:5:in `block in up'
/home/bob/src/alces-software/alces-flight-center/db/data/20180611161221_migrate_legacy_case_charges.rb:3:in `up'
```

We do have access to the User who initiated old additions of `credit_charges`, as [we're tracking this with `audited`](https://github.com/alces-software/alces-flight-center/blob/3f0c3b86f2a8edc4b5d7b0655ebcd229bbb72d48/app/models/case.rb#L38) and use to display this event in the case history, so I'm guessing we should be able to access and use this here?

Also possibly worth using the bang method here so we find out if creating the existing charges fails?

jamesremuscat replied:
That's actually revealed that `audited` isn't working for the new `CreditCharge`s (presumably because `credit_charge` isn't an extant column in the database).

In this case it's easier to grab the user from the audit of state (since the user who closes the case is also the one who added the charge) so I've done that.

Just need to fix the auditing of `CreditCharge`s now...!

jamesremuscat commented:
Rebased on `develop` to resolve yet more conflicts - have now addressed all of your points, @bobwhitelock :slightly_smiling_face: 

The data migration now removes the old audit entries for `credit_charge`, and I've added a new migration to remove the `legacy_credit_charge` column from `Case` (because with the new data migration gem we can do that).

bobwhitelock commented:
Thanks for addressing those final points - using `validates_associated` seems good, and good idea removing the `legacy_credit_charge` column now. I think this all looks good now, thanks for all your work on it :slightly_smiling_face: :shipit:
